### DISCUSSION TITLE
mysqlconnector issue with binary strings

### DISCUSSION THREAD

**Discussion by Awilen-Bernkastel** [sqlalchemy/sqlalchemy#7690] [Usage Questions] [answered]
I understand this is not a question, however I don't feel like this is such a broad issue that it warrants its own bug report, I've found an answer to my issue and wanted to share.

I just started working with sqlacodegen using sqlalchemy 1.4.31, and I've stumbled upon a TypeError issue in the method `MySQLTableDefinitionParser._describe_to_create` :

`TypeError: sequence item 2: expected str instance, bytes found`

This is the faulting line: `buffer.append(" ".join(line))`, line 306 in /dialects/mysql/reflection.py

This is the content of `line`: `[' ', 'col_name', b'int unsigned']` (col_name has back-quotes too, but this messes up the formatting.) Notice the `b''` string.

Going up the call stack, I found the method `Base._compat_fetchall` (from /dialects/mysql/base.py) that uses an instance of `_DecodingRow` to convert each row of `rp.fetchall()` (the list of columns) to UTF8, however this method is never called. Instead, the subclassed method `MySQLDialect_mysqlconnector._compat_fetchall` is called and doesn't do any conversion at all, instead returning `rp.fetchall()` as-is. The latter method is at line 214 of /dialects/mysql/mysqlconnector.py

Commenting the subclassed method `MySQLDialect_mysqlconnector._compat_fetchall` (and for consistency's sake the sister method `MySQLDialect_mysqlconnector._compat_fetchone`) allowed the execution of sqlacodegen without issue, using the method of the same name of the class Base instead.

Windows 10 20H2 19042.1415
Python 3.10.2
MySQL 8.0.27

**Comment by zzzeek** [comment, marked as answer]
hey there -

inconsistent "byte" results are probably the main reason we recommend against using mysqlconnector. it has a strange interface and has all kinds of areas where it inconsistently returns strings or bytes, it's not clear which of these issues are bugs in the driver and which are just part of its not very intuitive usage interface.  if you use the recommended drivers like pymysql / mysqlclient, there should not be any issue.

**Reply by Awilen-Bernkastel** [reply]
Thanks for the pointer! I tried looking for mysqlconnector in this Discussion section but nothing came up, so I figured I'd write one with as much info as possible. I'll give those drivers a try!

**Reply by zzzeek** [reply]
we have a "dragon" note regarding the driver here: https://docs.sqlalchemy.org/en/14/dialects/mysql.html#module-sqlalchemy.dialects.mysql.mysqlconnector

the driver is more mature now so we could probably get the rest of it working but in the past it hasn't been worth it, as the driver kept breaking on us back when we had it in CI, despite us going back a few times and refactoring the dialect for newer updates of the driver.
